ethers-providers/src/gas_oracle/mod.rs

```diff
@@ -0,0 +1,75 @@
+mod eth_gas_station;
+pub use eth_gas_station::EthGasStation;
+
+mod etherchain;
+pub use etherchain::Etherchain;
+
+mod etherscan;
+pub use etherscan::Etherscan;
+
+use ethers_core::types::U256;
+
+use async_trait::async_trait;
+use reqwest::Error as ReqwestError;
+use thiserror::Error;
+
+/// Various gas price categories. Choose one of the available
+#[derive(Debug)]
+pub enum GasCategory {
+    SafeLow,
+    Standard,
+    Fast,
+    Fastest,
+}
+
+#[derive(Error, Debug)]
+/// Error thrown when fetching data from the `GasOracle`
+pub enum GasOracleError {
+    /// An internal error in the HTTP request made from the underlying
+    /// gas oracle
+    #[error(transparent)]
+    HttpClientError(#[from] ReqwestError),
+
+    /// An internal error thrown when the required gas category is not
+    /// supported by the gas oracle API
+    #[error("gas category not supported")]
+    GasCategoryNotSupported,
+}
+
+/// `GasOracle` is a trait that an underlying gas oracle needs to implement.
+///
+/// # Example
+///
+/// ```no_run
+/// use ethers::providers::{
+///     gas_oracle::{EthGasStation, Etherscan, GasCategory, GasOracle},
+/// };
+///
+/// # async fn foo() -> Result<(), Box<dyn std::error::Error>> {
+/// let eth_gas_station_oracle = EthGasStation::new(Some("my-api-key"));
+/// let etherscan_oracle = EthGasStation::new(None).category(GasCategory::SafeLow);
+///
+/// let data_1 = eth_gas_station_oracle.fetch().await?;
+/// let data_2 = etherscan_oracle.fetch().await?;
+/// # Ok(())
+/// # }
+/// ```
+#[async_trait]
+pub trait GasOracle: std::fmt::Debug {
```

Comment:
Kind of similar comment here, you may want to be conservative to require something like
```
    : Send 
    + Sync 
    + 'static 
    + std::any::Any 
    + std::fmt::Debug
```

I'm not 100% sure how not specifying Send/Sync would interact with futures that take ownership on parameters that contain something like `Box<dyn GasOracle>` or similar.